### PR TITLE
feat: All AWS services emit traces

### PR DESCRIPTION
**Pull request by jterapin**
**Description**: Currently, only AWS services outlined [here](https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/main/instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/services.rb#L7) can emit tracing. This PR ensures that all AWS services are supported.

**Other changes**:
* Refactor `MessagingHelper` and the existing `Handler` in preparation to integrate the AWS SDK for Ruby Observability support.

### CODE REVIEW

File: instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/instrumentation.rb
Comment (alextwoods) [inline instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/instrumentation.rb:75]: Nit: the name is a little confusing - I think it should maybe be `loaded_service?`

File: instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/instrumentation.rb
Comment (alextwoods) [inline instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/instrumentation.rb:54]: Nit: I know this is an existing name, but I find it a little confusing.  I think this is internal only, so maybe we can rename as something like `loaded_service_clients`?

File: instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/instrumentation.rb
Comment (alextwoods) [inline instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/instrumentation.rb:62]: ditto on naming here, maybe `loaded_v3_clients` or something like that.

File: instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/instrumentation.rb
Comment (alextwoods) [inline instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/instrumentation.rb:58]: Is there a reason the dynamic approach doesn't work in V2?  I think it should and may be better to avoid checking against (and defining the current major version).

File: instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/messaging_helper.rb
Comment (alextwoods) [inline instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/messaging_helper.rb:48]: extra comment

File: instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/instrumentation.rb
Comment (jterapin) [reply instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/instrumentation.rb]: Thanks to CI (🥹✨) failing on V2 cases. I think my assumption that V2 Clients inherits from `Seahorse::Client::Base` was wrong. What it actually happens is that V2 Clients inherits from `Aws::Client` and THEN, inherits from `Seahorse::Client::Base`. So I have made a minor adjustment to cover this.

File: instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/handler.rb
Comment (jterapin) [reply instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/handler.rb:17]: In theory, I think all service clients should have an operation (and its name) on the `context` object (at least for V3). I'm not too sure about V2 since that has been long deprecated but I imagined it should be same since we use service models to generate the clients.

However, I will lean to leave this as it is - since I want to keep the logic the same much as it can to keep the legacy behavior before i introduce the integration with our recent support of Observability.

**Review by kaylareopelle — COMMENTED** [review]
I think this looks really good! Thank you so much for your contribution. I have one small question before approval.

File: instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/instrumentation.rb
Comment (jterapin) [reply instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/instrumentation.rb:74]: Hi! That's a great question. I think it's probably worthwhile for me to update this method with some comments of context.

Yeah, basically - we want to use this method to be compatible with MV 2 and 3 of the Ruby SDK. Both of these two MVs works a bit differently on how service clients are created underneath. V3 service client's superclass is `Seahorse::Client::Base` whereas V2 service client's superclass is `Aws::Client` - hence the conditional.

Comment (kaylareopelle) [reply instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/instrumentation.rb:74]: Ahh, okay. Thank you! The comments you added are helpful!

I believe the current condition is evaluated like:
`((service_module.const_get(:Client).superclass == Seahorse::Client::Base) || Aws::Client)`

so whenever `Seahorse::Client::Base` is not the client superclass, the `Aws::Client` constant is returned instead of a boolean. `Aws::Client` would also be returned for constants without the right superclass. 

For example:
```ruby
def test(var)
  var == 'b' || 'a'
end

test('b')
=> true
test('a')
=> 'a'
test('c')
=> 'a'
```

I think what you may want is: 
`service_module.const_get(:Client).superclass == Seahorse::Client::Base || service_module.const_get(:Client).superclass == Aws::Client`

or maybe something like:
`[Seahorse::Client::Base, Aws::Client].include?(service_module.const_get(:Client).superclass)`

Comment (jterapin) [reply instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/instrumentation.rb:74]: Ahhh, thanks for this catch. You are correct. 😆
I'll push a change right now!

Comment (kaylareopelle) [reply instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/instrumentation.rb]: Thank you! Change looks great, though I think something else might need to be updated for the tests? 
https://github.com/open-telemetry/opentelemetry-ruby-contrib/actions/runs/10838057325/job/30082448049?pr=1150#step:3:575

Comment (jterapin) [reply instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/instrumentation.rb]: Ahh, I see what happened here (it was due to how V3 vs V2 works again 😔). 
Hopefully, my update did the trick.

**Review by kaylareopelle — APPROVED** [review]
Thanks, @jterapin! 🎉

**Review by Theghost256 — APPROVED** [review]

**Pull request merged**
